base/accumulate.jl

```diff
@@ -0,0 +1,566 @@
+    s, sstate = next(src, sstate0)
+    while true
+        @inbounds dest[i] = s
+
+        sdone = done(src, sstate)
+        ddone = done(destiter, dstate)
+
+        sdone && ddone && return dest
+        (sdone || ddone) && throw(DimensionMismatch("input and output array sizes and indices must match"))
+
+        i, dstate = next(destiter, dstate)
+        s, sstate = next(src, sstate)
+    end
+end
+
+
+"""
+    accumulate(op[, v0], A, dim::Integer)
+
+Cumulative operation `op` along the dimension `dim`. See also
+[`accumulate!`](@ref) to use a preallocated output array. For common operations
+there are specialized variants of `accumulate`, see:
+[`cumsum`](@ref), [`cumprod`](@ref)
+
+# Examples
+```jldoctest
+julia> accumulate(+, fill(1, 3, 3), 1)
+3×3 Array{Int64,2}:
+ 1  1  1
+ 2  2  2
+ 3  3  3
+
+julia> accumulate(+, fill(1, 3, 3), 2)
+3×3 Array{Int64,2}:
+ 1  2  3
+ 1  2  3
+ 1  2  3
+```
+"""
+accumulate(op, X, dim::Int) = accumulate(op, uninitialized, X, dim)
+
+# split into (head, slice, tail) dimension iterators
+@inline function split_dimensions(X,dim::Integer)
+    inds = axes(X)
+    if dim > length(inds)
+        (CartesianIndices(inds), CartesianIndices(), CartesianIndices())
+    else
+        (CartesianIndices(inds[1:dim-1]), inds[dim], CartesianIndices(inds[dim+1:end]))
+    end
+end
+
+function accumulate(op, v0, X, dim::Integer)
+    dim > 0 || throw(ArgumentError("dim must be a positive integer"))
+    if length(X) == 0
+        # fallback on collect machinery
+        return collect(Accumulate(op, v0, X))
+    end
+    indH, indD, indT = split_dimensions(X,dim)
+
+    # TODO: this could be much nicer with the new iteration protocol
+    # unroll loops to get first element
+    sH = start(indH)
+    @assert !done(indH, sH)
+    iH,sH = next(indH, sH)
+
+    sD = start(indD)
+    @assert !done(indD, sD)
+    iD,sD = next(indD, sD)
+    pD = iD
+
+    sT = start(indT)
+    @assert !done(indT, sT)
+    iT,sT = next(indT, sT)
+
+    # first element
+    accv = reduce_first(op, v0, X[iH, iD, iT])
+    dest = similar(X, typeof(accv))
+    i = first(linearindices(dest))
+    dest[i] = accv
+
+    return _accumulate!(dest, i, op, v0, X, indH, indD, indT, sH, sD, sT, iH, iD, iT, pD)
+end
+
+
+"""
+    accumulate!(dest, op, A, dim::Integer)
+
+Cumulative operation `op` on `A` along the dimension `dim`, storing the result in `B`.
+See also [`accumulate`](@ref).
+
+# Examples
+```jldoctest
+julia> A = [1 2; 3 4];
+
+julia> B = [0 0; 0 0];
+
+julia> accumulate!(-, B, A, 1);
+
+julia> B
+2×2 Array{Int64,2}:
+  1   2
+ -2  -2
+
+julia> accumulate!(-, B, A, 2);
+
+julia> B
+2×2 Array{Int64,2}:
+ 1  -1
+ 3  -1
+```
+"""
+accumulate!(dest, op, X, dim::Integer) = accumulate!(dest, op, uninitialized, X, dim::Integer)
+
+function accumulate!(dest, op, v0, X, dim::Integer)
+    dim > 0 || throw(ArgumentError("dim must be a positive integer"))
+    axes(A) == axes(B) || throw(DimensionMismatch("shape of source and destination must match"))
```

Comment:
What is `A` and `B` should be `dest` and `X`.

Reply:
I think we also need more `accumulate!` tests. Historically `accumulate!` would be implicitly called anyway by each `accumulate` test so this was not necessary. AFAICT there is only a single `accumulate!` test now.